backends/platform/ios7/ios7_video_metal.mm

```diff
@@ -0,0 +1,292 @@
+- (void)layoutSublayersOfLayer:(CAMetalLayer *)layer {
+	if (layer == self.layer) {
+		[self addEvent:InternalEvent(kInputScreenChanged, 0, 0)];
+	}
+	[super layoutSublayersOfLayer:layer];
+}
+
+- (id)initWithFrame:(struct CGRect)frame andMetalDevice:(id<MTLDevice>_Nonnull)device {
+	self = [super initWithFrame:frame];
+
+	_metalDevice = device;
+
+	// Initialise the Metal context
+	[self createMetalContext];
+
+	// Initialize the OpenGL ES context
+	[self createOpenGLESContext];
+
+	return self;
+}
+
+- (void)createMetalContext {
+	_metalLayer = (CAMetalLayer *)self.layer;
+	_metalLayer.framebufferOnly = NO;
+	_metalLayer.pixelFormat = MTLPixelFormatRGBA8Unorm;
+
+	_commandQueue = [_metalDevice newCommandQueue];
+	[_metalLayer setDevice:_metalDevice];
+	[_metalLayer setDrawableSize:CGSizeMake(self.frame.size.width * self.contentScaleFactor, self.frame.size.height * self.contentScaleFactor)];
+
+	_metalLibrary = [_metalDevice newLibraryWithSource:[NSString stringWithUTF8String:kernelFunction] options:nil error:nil];
+	_kernelFunctionFlipY = [_metalLibrary newFunctionWithName:@"flip_y"];
+}
+
+- (void)createOpenGLESContext {
+	_mainContext = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2];
+
+	// In case creating the OpenGL ES context failed, we will error out here.
+	if (_mainContext == nil) {
+		printError("Could not create OpenGL ES context.");
+		abort();
+	}
+
+	// main thread will always use _mainContext
+	[EAGLContext setCurrentContext:_mainContext];
+}
+
+- (uint)createOpenGLContext {
+	// Create OpenGL context with the sharegroup from the context
+	// connected to the Apple Core Animation layer
+	if (!_openGLContext && _mainContext) {
+		_openGLContext = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2 sharegroup:_mainContext.sharegroup];
+
+		if (_openGLContext == nil) {
+			printError("Could not create OpenGL ES context using sharegroup");
+			abort();
+		}
+		// background thread will always use _openGLContext
+		if ([EAGLContext setCurrentContext:_openGLContext]) {
+			[self setupRenderBuffer];
+		}
+	}
+	return CVOpenGLESTextureGetName(_openGLTextureRef);
+}
+
+- (void)destroyOpenGLContext {
+	[_openGLContext release];
+	_openGLContext = nil;
+}
+
+- (void)setupOpenGLTextureCache {
+	if (_openGLTextureCache == nullptr) {
+		CVOpenGLESTextureCacheCreate(
+			kCFAllocatorDefault,
+			NULL, _openGLContext,
+			NULL, &_openGLTextureCache);
+	}
+
+	CFDictionaryRef emptyDict = CFDictionaryCreate(
+		kCFAllocatorDefault,
+		NULL, NULL, 0,
+		&kCFTypeDictionaryKeyCallBacks,
+		&kCFTypeDictionaryValueCallBacks);
+
+	CFMutableDictionaryRef attributes = CFDictionaryCreateMutable(
+		kCFAllocatorDefault, 1,
+		&kCFTypeDictionaryKeyCallBacks,
+		&kCFTypeDictionaryValueCallBacks);
+
+	CFDictionarySetValue(
+		attributes,
+		kCVPixelBufferIOSurfacePropertiesKey, // This creates a shared IOSSurface
+		emptyDict);
+
+	if (_sharedPixelBuffer != nullptr) {
+		CFRelease(IOSurfaceRef(_sharedPixelBuffer));
```

Comment:
Documentation of CVPixelBufferCreate says to use CVPixelBufferRelease to release.